Solution for #527

sjaakd opened the pull request:
The solution introduces a fix for supporting wildcard extends bounds on the source side and wildcard super bounds on the target side.

Next to this, error handling is introduced for the reverse case and for all typevar cases.

Selection of wildcard methods needs to be revisited (see #527). I'll open a new issue for this as soon as this one is completed.

sjaakd commented on processor/src/main/java/org/mapstruct/ap/model/common/Type.java:
```diff
@@ -665,4 +690,57 @@ public int compareTo(Type o) {
+     */
+    public Type getTypeBound() {
+
+        Type boundingBase = this;
```
yes it would. :smile:. I'll fix that. I initially thought of adding this in the `TypeFactory`. Anyway, I now fixed the `IterableMapping` and `MapMapping` but perhaps it should be added in the `PropertyMapping` as well (separate PR, don't want to make this too big).

agudian commented on processor/src/test/java/org/mapstruct/ap/test/collection/wildcard/IterableTypeVarBoundMapperOnMethod.java:
```diff
@@ -0,0 +1,32 @@
+@Mapper
+public interface IterableTypeVarBoundMapperOnMethod {
+
+    <T extends String> List<T> map(List<String> in);
```
Funny that this even compiles, as String is final already... I could imagine this to break in future comipler versions, though...

sjaakd replied:
Of course.. Needed a simple type but forgot that a string is final. I'll change that..

agudian commented:
OK, I took a closer look and I didn't find anything wrong with it... (I didn't wrap my head around it fully, so I can't say I completely understood everything of it). :+1: 

Just in case gunnar and you can spin a review on #84 this weekend, would it be possible to hold back the merge until after #84 lands on master? I'm not exactly sure if I might have to adapt something there as well, if this PR would be merged first (my guess: no).

sjaakd commented:
> Just in case gunnar and you can spin a review on #84 this weekend, would it be possible to hold back the merge until after #84 lands on master? I'm not exactly sure if I might have to adapt something there as well, if this PR would be merged first (my guess: no).

Sure, no problem.


I've now fixed this for Itterable and Map mappings. But I need to look into Property mappings as well (at least add a test case).

gunnarmorling commented:
Wow, that's a huge one :) Could you maybe describe a bit on a high level what has changed? Many thanks!

gunnarmorling commented on processor/src/test/java/org/mapstruct/ap/test/collection/wildcard/GoodIdea.java:
```diff
@@ -0,0 +1,49 @@
+ *
+ * @author Sjaak Derksen
+ */
+public class GoodIdea {
```
:)

gunnarmorling commented on processor/src/main/java/org/mapstruct/ap/util/Message.java:
```diff
@@ -88,6 +88,12 @@
     RETRIEVAL_PRIMITIVE_RETURN( "Can't generate mapping method with primitive return type." ),
     RETRIEVAL_ENUM_TO_NON_ENUM( "Can't generate mapping method from enum type to non-enum type." ),
     RETRIEVAL_NON_ENUM_TO_ENUM( "Can't generate mapping method from non-enum type to enum type." ),
+    RETRIEVAL_TYPE_VAR_SOURCE( "Can't generate mapping method for a generic type variable source." ),
```
What is a "generic type variable source"? Is it an unbound wildcard?

sjaakd replied:
generics come basically in 2 forms: parameterized (like <T extends Something> on method / class level) and wildcards (the question mark). A generic type variable is the first one.  Checkout http://docs.oracle.com/javase/7/docs/api/javax/lang/model/type/TypeKind.html

gunnarmorling commented:
From a quick view I think it's fine, still a quick explanation of what's happening in this change might be helpful.

Btw. did you guys notice we crossed the 100 stargazers mark. Yeah :)

sjaakd commented:
> From a quick view I think it's fine, still a quick explanation of what's happening in this change might be helpful.

I'll add an explanation tomorrow.



> Btw. did you guys notice we crossed the 100 stargazers mark. Yeah :)

Cool! :smile:

sjaakd commented:
After the explanation, and after the merge of @agudian's piece-de-resistance :smile:, can I squash and merge this one?

sjaakd commented:
Rebased and squashed into one commit.

This PR introduces selection and forging of wildcard parameterized Iterable and Map mapping methods. For target elements, a super bound is supported. For source elements an extends bound.

For the first, the forged (and selected) method looks like this:
```java
    protected List<? super org.mapstruct.ap.test.collection.wildcard.Plan> ideaListToPlanList(List<Idea> list) {
        if ( list == null ) {
            return null;
        }

        List<? super org.mapstruct.ap.test.collection.wildcard.Plan> list_ = new ArrayList<Plan>();

        for ( Idea idea : list ) {
            list_.add( map( idea ) );
        }

        return list_;
    }
```
 for the latter the forged (and selected) method looks like this:
```java
   protected List<Plan> ideaListToPlanList(List<? extends org.mapstruct.ap.test.collection.wildcard.Idea> list) {
        if ( list == null ) {
            return null;
        }

        List<Plan> list_ = new ArrayList<Plan>();

        for ( Idea idea : list ) {
            list_.add( map( idea ) );
        }

        return list_;
    }
```
The same goes for map mapping methods.

Note: It seems that the used `TypeMirror` api uses fully qualified names when representing wild card types. This is visible in the generated code and can be improved in the future.

---------------------

Also some code and tests have been added to prevent selection (and forging) of typevar parameterized types. An appropriate error will be generated. Examples:


```java
@Mapper
public interface IterableTypeVarBoundMapperOnMapper<T extends BigDecimal> {

     List<BigDecimal> map(List<T> in);
}
```
and
```java
@Mapper
public interface IterableTypeVarBoundMapperOnMethod {

    <T extends BigDecimal> List<T> map(List<BigDecimal> in);
}
```

agudian commented:
I'm just seeing it now with your example output above... 

> Note: It seems that the used TypeMirror api uses fully qualified names when representing wild card types. This is visible in the generated code and can be improved in the future.

Did you try it out with a different annotation processor implementation, e.g. with Eclipse? A simple addition to one of the integration tests could do that for you, if you don't want to pull out your dusty Eclipse installation... ;-). Just to make sure that this still produces compilable code...

sjaakd commented:
> Did you try it out with a different annotation processor implementation, e.g. with Eclipse? A simple addition to one of the integration tests could do that for you, if you don't want to pull out your dusty Eclipse installation... ;-). Just to make sure that this still produces compilable code...

I'll add an integration test. Do you suspect that other annotations processor will handle this differently?

sjaakd commented:
> Meh, I don't know... just a feeling. We had something vaguely similar with the string representation of executable elements, remember?

Ok.. Added some integration test.. No problem.. Will proceed to merge :smile: . Thanks.
